tests/frontend/test_program.py

```diff
@@ -0,0 +1,308 @@
+            ops.Dgate(0.5) | q[0]
+        # now there is one gate in the queue
+        assert len(prog) == 1
+
+        with prog.context as q:
+            ops.BSgate(0.5, 0.3) | (q[1], q[0])
+        assert len(prog) == 2
+
+    def test_identity_command(self, prog):
+        """Tests that the None command acts as the identity"""
+        identity = program.Command(None, prog.register[0])
+        prog.circuit.append(identity)
+        assert len(prog) == 1
+        prog = prog.compile("base")
+        assert len(prog) == 0
+
+    def test_parent_program(self):
+        """Continuing one program with another."""
+        D = ops.Dgate(0.5)
+        prog = sf.Program(3)
+        with prog.context as q:
+            D | q[1]
+            ops.Del | q[0]
+        cont = sf.Program(prog)
+        with cont.context as q:
+            D | q[0]
+            r = ops.New(1)
+            D | r
+        assert cont.can_follow(prog)
+        assert prog.reg_refs == cont.init_reg_refs
+        assert prog.unused_indices == cont.init_unused_indices
+
+    def test_print_commands(self, eng, prog):
+        """Program.print and Engine.print_applied return correct strings."""
+        prog = sf.Program(2)
+        res = []
+        print_fn = lambda x: res.append(x.__str__())
+
+        # prog should now be empty
+        prog.print(print_fn)
+        assert res == []
+
+        # define some gates
+        D = ops.Dgate(0.5)
+        BS = ops.BSgate(2 * np.pi, np.pi / 2)
+        R = ops.Rgate(np.pi)
+
+        with prog.context as q:
+            alice, bob = q
+            D | alice
+            BS | (alice, bob)
+            ops.Del | alice
+            R | bob
+            charlie, = ops.New(1)
+            BS | (bob, charlie)
+            ops.MeasureX | bob
+            ops.Dgate(bob).H | charlie
+            ops.Del | bob
+            ops.MeasureX | charlie
+
+        res = []
+        prog.print(print_fn)
+
+        expected = [
+            "Dgate(0.5, 0) | (q[0])",
+            "BSgate(6.283, 1.571) | (q[0], q[1])",
+            "Del | (q[0])",
+            "Rgate(3.142) | (q[1])",
+            "New(1)",
+            "BSgate(6.283, 1.571) | (q[1], q[2])",
+            "MeasureX | (q[1])",
+            "Dgate(RR(q[1]), 0).H | (q[2])",
+            "Del | (q[1])",
+            "MeasureX | (q[2])",
+        ]
+
+        assert res == expected
+
+        state = eng.run(
+            prog, compile=False
+        )  # FIXME optimization can change gate order, however this is not a good way of avoiding it
+        res = []
+        eng.print_applied(print_fn)
+        assert res == ["Run 0:"] + expected
+
+
+class TestRegRefs:
+    """Testing register references."""
+
+    def test_corrupt_index(self, prog):
+        """User messes up the RegRef indices."""
+        with prog.context as q:
+            q[0].ind = 1
+            with pytest.raises(
+                program.RegRefError, match="RegRef state has become inconsistent"
+            ):
```

Comment:
Could we increase the allowed line length to 100--120 chars to prevent black splitting up lines like this?

Reply:
I mean this is not the worst formatting but I tend to use slightly longer lines myself...

Reply:
We should pick a standard Black line length to ensure that we are all using the same formatting, @co9olguy @trbromley @jswinarton

Reply:
go ahead and just pick one

Reply:
Can I make an executive decision and say 100 chars per line